core/sr-primitives/src/generic/unchecked_extrinsic.rs

```diff
@@ -132,52 +150,70 @@ impl<Address: Codec, Index: HasCompact + Codec, Signature: Codec, Call: Decode>
 		Some(UncheckedExtrinsic {
 			signature: Decode::decode(input)?,
 			function: Decode::decode(input)?,
+			tip: Decode::decode(input).unwrap_or_default(),
```

Comment:
Can we add some tests to make sure we maintain backward compatibility with previous encoding?

Reply:
Also it seems that you can add any junk to the transaction after `function` and we will just ignore it. We need to triple check that the thing that actually ends up in the blockchain is not the original `input` that user sent, but rather a re-encoding of decoded data to make sure the junk doesn't end up there.

Reply:
There is a already a test for this: 
https://github.com/paritytech/substrate/pull/2930/files/b110af6a218b6d84d4009ec93a5d174890330d59#diff-d99c5e275f7eef692ff59aa71ed00956R449

BUT: what you say about the junk being decoded is true but I see that as merely the _price of being backward compatible with old transactions that don't have a tip in them_. Tho' if the codec allowed us to peek the input here and do this more efficiently, that'd be better I suppose. 

But this still has the flaw of potential spamming: anyone sending a --even--__valid__ transaction can just post-fix it with some random crap bytes and we _waste a little bit of time trying to decode it_, then falling back to the `_or_default()` once it fails. Again, I think someone with codec knowledge might be able to propose a better solution here than me. 

> We need to triple check that the thing that actually ends up in the blockchain is not the original input that user sent, but rather a re-encoding of decoded data to make sure the junk doesn't end up there.

I don't quite get what you mean but a junk will for sure not land in the blockchain now (at least I don't see how). it might just waste a bit of our time. Note that the default impl for `Tip` is `None` so this basically tries to say: _Decode the tip if it exists or `None` if not_.

Reply:
cc @bkchr

Reply:
> I don't quite get what you mean but a junk will for sure not land in the blockchain now (at least I don't see how).

Well, that's why I asked if we do `Decode` the payload and then `Encode` it before storing. I imagine that as a straightfoward optimization we could decide to avoid re-encoding and just store the same payload that we got which would cause issues. But If we don't do that and don't plan to - it's totally fine.